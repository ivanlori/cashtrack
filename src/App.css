.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.action-button {
  @apply flex;
  @apply items-center;
  @apply justify-center;
  @apply w-12;
  @apply h-12;
  @apply transition-colors;
  @apply duration-150;
  @apply ease-in-out;
  @apply bg-transparent;
  @apply border;
  @apply border-solid;
  @apply shadow-none;
  @apply cursor-pointer;
  @apply rounded-2xl;
  @apply border-stone-300;
  @apply hover:bg-slate-100;
  @apply focus:bg-slate-100;
}