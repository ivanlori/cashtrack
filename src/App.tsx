import { ReactElement, useMemo, useState } from 'react';
import styles from './App.module.css';
import { useLiveQuery } from "dexie-react-hooks";
import moment from 'moment';
import cn from 'classnames';
import {
	LeadingActions,
	SwipeableList,
	SwipeableListItem,
	SwipeAction,
	TrailingActions,
	Type,
} from 'react-swipeable-list';
import 'react-swipeable-list/dist/styles.css';
import { CiCircleMinus, CiEdit } from "react-icons/ci";
import "dexie-export-import";
import { Controller, SubmitHandler, useForm } from 'react-hook-form';
import { Navbar } from './Navbar';
import { db } from './db';
import { CustomSelect, Option } from './components/Select';

type FormValues = {
	expense: string
	category: string
}

type IDataSaved = {
	id: string,
	value: string,
	date: Date,
	category: string
}

const months = [
	'Gennaio',
	'Febbraio',
	'Marzo',
	'Aprile',
	'Maggio',
	'Giugno',
	'Luglio',
	'Agosto',
	'Settembre',
	'Ottobre',
	'Novembre',
	'Dicembre'
]

const App = (): ReactElement => {
	const defaultValues: FormValues = useMemo(() => ({
		expense: '',
		category: ''
	}), [])

	const {
		register,
		handleSubmit,
		reset,
		control,
		formState: {
			errors
		},
	} = useForm<FormValues>({ defaultValues })
	const expenseTable = db.table('expenses')
	const [isLoadingList, setLoadingList] = useState(true)
	const [isLoadingPreviousTotal, setLoadingPreviousTotal] = useState(true)
	const [isLoadingCurrentTotal, setLoadingCurrentTotal] = useState(true)
	const [selectedCategory, setSelectedCategory] = useState<string>('')
	const [hasSubmitted, setHasSubmitted] = useState<boolean>(false)

	const onSubmit: SubmitHandler<FormValues> = async (data): Promise<void> => {
		await expenseTable.add({
			value: data.expense,
			date: new Date(),
			category: selectedCategory
		})
		setHasSubmitted(true)

		// works only for the expense field
		reset()
	}

	const allExpenses = useLiveQuery(async (): Promise<IDataSaved[]> => {
		setLoadingList(true)
		return await expenseTable
			.orderBy('date')
			.reverse()
			.toArray()
			.finally(() => setLoadingList(false))
	}, []);

	const totalPreviousMonthExpenses = useLiveQuery(async (): Promise<number> => {
		setLoadingPreviousTotal(true)
		const previousMonth = new Date().getMonth()
		let total = 0

		return await expenseTable
			.filter((row) => new Date(row.date).getMonth() + 1 === previousMonth)
			.each((item: IDataSaved) => total += Number(item.value))
			.then(() => Number(total.toFixed(2)))
			.finally(() => setLoadingPreviousTotal(false))
	}, [])

	const totalCurrentMonthExpenses = useLiveQuery(async (): Promise<number> => {
		setLoadingCurrentTotal(true)
		const currentMonth = new Date().getMonth() + 1
		let total = 0

		return await expenseTable
			.filter((row) => new Date(row.date).getMonth() + 1 === currentMonth)
			.each((item: IDataSaved) => total += Number(item.value))
			.then(() => Number(total.toFixed(2)))
			.finally(() => setLoadingCurrentTotal(false))
	}, []);

	const deleteItem = (id: string): void => {
		expenseTable.delete(id)
	}

	const showEditInput = (id: string, valueToEdit: string): void => {
		const newValue = prompt('Inserisci il nuovo valore', valueToEdit)
		if (newValue) {
			expenseTable.update(id, { value: newValue })
		}
	}

	const getFormattedTotal = (total: number | undefined): string => (
		total ? `${total > 0 ? '-' : ''}${total}€` : 'N/D'
	)

	const renderLoading = (width: string): ReactElement => (
		<span className={`${width} block h-6 bg-gray-300 rounded animate-pulse`} />
	)

	return (
		<div className="App">
			<Navbar />
			<main>
				<div className="mx-10 mb-3">
					<div className="previous-month flex items-center gap-2">
						<span className="text-md text-gray-400">
							{months[new Date().getMonth() - 1]}:
						</span>
						{' '}
						{isLoadingPreviousTotal ? renderLoading('w-14') : (
							<span className="text-xl text-red-400">
								{getFormattedTotal(totalPreviousMonthExpenses)}
							</span>
						)}
					</div>
					<div className="current-month flex items-center gap-2">
						<span className="text-xl text-gray-600">Mese corrente:</span>
						{' '}
						{isLoadingCurrentTotal ? renderLoading('w-14') : (
							<span className="text-2xl text-red-600">
								{getFormattedTotal(totalCurrentMonthExpenses)}
							</span>
						)}
					</div>
				</div>
				<form
					onSubmit={handleSubmit(onSubmit)}
					className="h-60 mx-10"
				>
					<div className="flex flex-col gap-3">
						<div className="flex relative items-center">
							<input
								className={cn(styles.Input, {
									"border-red-600": errors.expense?.message
								})}
								type="number"
								step=".01"
								placeholder="0"
								{...register('expense', {
									onChange: () => setHasSubmitted(false),
									pattern: {
										value: /^\d+(?:[.,]\d+)*$/,
										message: 'Valore non valido'
									},
									required: {
										value: true,
										message: 'Inserisci la spesa'
									}
								})}
							/>
							<span className="absolute right-0 mr-3 text-xl">€</span>
						</div>
						<Controller
							control={control}
							name="category"
							rules={{
								required: {
									value: true,
									message: 'Inserisci la categoria'
								},
							}}
							render={({ field: { onChange } }): ReactElement => (
								<CustomSelect
									showErrorStyle={errors.category?.message !== undefined}
									resetValue={hasSubmitted}
									onChange={(value: Option | null): void => {
										setSelectedCategory(value?.label || '')
										setHasSubmitted(false)
										onChange(value?.value)
									}}
								/>
							)}
						/>
					</div>
					<button
						type="submit"
						className="bg-blue-500 rounded-lg text-xl mt-3 w-full text-white text-center px-5 py-3 transition duration-300 ease-in-out hover:bg-blue-600"
					>
						Aggiungi
					</button>
					<span className="text-red-600 mt-2 block">
						{errors.expense?.message}
						<br />
						{errors.category?.message}
					</span>
				</form>
				<SwipeableList className="mt-5 mx-auto" Tag="ul" type={Type.IOS}>
					{allExpenses?.map((item: IDataSaved) => {
						return (
							<SwipeableListItem
								key={item.id}
								maxSwipe={0.5}
								leadingActions={
									(<LeadingActions>
										<SwipeAction
											onClick={() => showEditInput(item.id, item.value)}
										>
											<CiEdit />
										</SwipeAction>
									</LeadingActions>)
								}
								trailingActions={
									(<TrailingActions>
										<SwipeAction
											onClick={() => deleteItem(item.id)}
										>
											<CiCircleMinus />
										</SwipeAction>
									</TrailingActions>)
								}
								className={styles.Item}
							>
								<div className="flex flex-col items-end gap mx-10 py-2">
									<div>
										{
											isLoadingList ? renderLoading('w-40') : (
												<>
													<span className="text-gray-600 text-xl">
														{item.category}
													</span>
													{' '}
													<span className="text-red-600">
														{`-${item.value}€`}
													</span>
												</>
											)
										}
									</div>
									{isLoadingList ? renderLoading('w-40') : (
										<span className="text-base text-gray-600">
											{moment(item.date).format('LLL')}
										</span>
									)}
								</div>
							</SwipeableListItem>
						)
					})}
				</SwipeableList>
			</main>
		</div>
	);
}

export default App;
