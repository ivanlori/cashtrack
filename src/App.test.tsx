import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import selectEvent from 'react-select-event';
import App from './App';

// Polyfill for structuredClone
if (typeof structuredClone === 'undefined') {
	global.structuredClone = (obj: any) => JSON.parse(JSON.stringify(obj));
}

test('should add a price item', async () => {
	render(
		<BrowserRouter>
			<App />
		</BrowserRouter>
	);
	const input = screen.getByTestId('expense-input')
	const categorySelect = screen.getByLabelText('Seleziona o crea una categoria')
	const addBtn = screen.getByText('Aggiungi')

	await userEvent.type(input, '100')
	await waitFor(() => {
		selectEvent.create(categorySelect, 'Food')
	})
	await waitFor(() => {
		selectEvent.select(categorySelect, 'Crea categoria "Food"')
	})
	await userEvent.click(addBtn)

	await waitFor(() => {
		expect(screen.getByTestId('expense-item')).toHaveTextContent('Food -100â‚¬')
	})
})
