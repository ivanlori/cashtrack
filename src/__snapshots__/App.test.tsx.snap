// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <header
        class="sticky bg-white top-0 mb-6 z-50"
      >
        <div
          class="px-5"
        >
          <div
            class="p-4 flex justify-between"
          >
            <div
              class="flex justify-center items-center gap-2"
            >
              <span
                class="my-0 flex text-dark font-semibold text-[1.35rem]/[1.2] flex-col justify-center"
              >
                Budget
              </span>
            </div>
            <div
              class="flex items-center"
            >
              <div
                class="relative flex items-center gap-3"
              >
                <a
                  class="ActionButton"
                  href="/settings"
                  role="button"
                >
                  <svg
                    fill="currentColor"
                    height="25"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="25"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      id="Settings"
                    >
                      <g>
                        <path
                          d="M12.6,20.936H11.3a.883.883,0,0,1-.852-.654l-.774-2.833-2.5,1.435a.886.886,0,0,1-1.06-.138l-.925-.919a.884.884,0,0,1-.143-1.066l1.469-2.545L6.509,14.2l-2.787-.747a.882.882,0,0,1-.654-.851V11.3a.882.882,0,0,1,.652-.85l2.839-.777L5.12,7.171a.885.885,0,0,1,.141-1.062l.918-.918A.885.885,0,0,1,7.24,5.049L9.792,6.514l.012,0,.745-2.79a.881.881,0,0,1,.851-.655h1.3a.883.883,0,0,1,.852.655l.762,2.838,2.509-1.441a.885.885,0,0,1,1.059.138l.926.919a.882.882,0,0,1,.141,1.067L17.483,9.777l.008.022,2.786.746a.883.883,0,0,1,.653.851v1.3a.883.883,0,0,1-.654.852l-2.837.774,1.439,2.505a.881.881,0,0,1-.141,1.063l-.917.917a.888.888,0,0,1-1.063.141l-2.539-1.462L14.2,17.5l-.745,2.785A.885.885,0,0,1,12.6,20.936Zm-1.21-1h1.119l.738-2.756a.888.888,0,0,1,.528-.592l.134-.052a.873.873,0,0,1,.76.057l2.51,1.445.789-.789-1.423-2.478a.881.881,0,0,1-.048-.78l.052-.125a.875.875,0,0,1,.584-.51l2.8-.749v-1.12l-2.755-.737a.885.885,0,0,1-.592-.529l-.052-.132a.882.882,0,0,1,.057-.763L18.04,6.818l-.8-.79-2.48,1.425a.878.878,0,0,1-.772.052l-.115-.047a.888.888,0,0,1-.518-.588l-.748-2.806H11.492l-.738,2.762a.883.883,0,0,1-.539.6l-.12.045a.874.874,0,0,1-.751-.058L6.822,5.962l-.789.789L7.455,9.227a.886.886,0,0,1,.046.785l-.051.12a.876.876,0,0,1-.579.5l-2.8.758v1.121l2.757.738a.889.889,0,0,1,.591.525l.048.121a.874.874,0,0,1-.055.77L5.958,17.181l.8.791,2.47-1.419a.878.878,0,0,1,.787-.045l.106.044a.874.874,0,0,1,.526.591ZM9.75,17.482l.008,0ZM9.6,17.421l.007,0ZM6.487,14.147h0Zm.044-4.411h0Zm7.724-3.2Z"
                        />
                        <path
                          d="M12,15a3,3,0,1,1,3-3A3,3,0,0,1,12,15Zm0-5a2,2,0,1,0,2,2A2,2,0,0,0,12,10Z"
                        />
                      </g>
                    </g>
                  </svg>
                </a>
              </div>
            </div>
          </div>
        </div>
      </header>
      <main>
        <div
          class="mx-10 mb-3"
        >
          <div
            class="previous-month flex items-center gap-2"
          >
            <span
              class="text-md text-gray-400"
            >
              Maggio
              :
            </span>
             
            <span
              class="w-14 block h-6 bg-gray-300 rounded animate-pulse"
            />
          </div>
          <div
            class="flex items-center gap-2"
          >
            <div
              class="current-month gap-2 flex items-center cursor-pointer"
            >
              <span
                class="text-xl text-gray-600"
              >
                Mese corrente:
              </span>
              <span
                class="w-14 block h-6 bg-gray-300 rounded animate-pulse"
              />
            </div>
          </div>
          <div
            class="transition-all duration-300 ease-in-out max-h-0 invisible"
          >
            <ul
              class="flex flex-wrap gap-4 p-3"
            />
          </div>
        </div>
        <form
          class="h-60 mx-10"
        >
          <div
            class="flex flex-col gap-3"
          >
            <div
              class="flex relative items-center"
            >
              <input
                class="Input"
                name="expense"
                placeholder="0"
                step=".01"
                type="number"
              />
              <span
                class="absolute right-0 mr-3 text-xl"
              >
                â‚¬
              </span>
            </div>
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                class=" css-18xrgii-control"
              >
                <div
                  class=" css-1fdsijx-ValueContainer"
                >
                  <div
                    class=" css-1jqq78o-placeholder"
                    id="react-select-2-placeholder"
                  >
                    Seleziona o crea una categoria
                  </div>
                  <div
                    class=" css-qbdosj-Input"
                    data-value=""
                  >
                    <input
                      aria-activedescendant=""
                      aria-autocomplete="list"
                      aria-describedby="react-select-2-placeholder"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-2-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-1xc3v61-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            class="bg-blue-500 rounded-lg text-xl mt-3 w-full text-white text-center px-5 py-3 transition duration-300 ease-in-out hover:bg-blue-600"
            type="submit"
          >
            Aggiungi
          </button>
          <span
            class="text-red-600 mt-2 block"
          >
            <br />
          </span>
        </form>
        <ul
          class="swipeable-list mt-5 mx-auto"
        />
      </main>
    </div>
  </div>
</body>
`;
